### PR TITLE
Removing confusion in why PR is being closed

### PR DESCRIPTION
**Pull request by riyasingh1004**
## Explanation
This PR removes the ambiguity  in invalid pull request branch names and resolves the issue #147

## Checklist
- [ ] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [ ] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by riyasingh1004** [comment]
@jameesjohn Can you kindly review the changes I have made.

File: lib/checkPullRequestBranch.js [hunk 1]
```diff
@@ -27,13 +27,43 @@ const checkBranch = async (context) => {
     return branchName.startsWith(prefix);
   });
 
-  if (branchIsInvalid) {
+  if(branchName=='develop')
+  {
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:31]:
```suggestion
  if (branchName === 'develop') {
```

File: lib/checkPullRequestBranch.js [hunk 2]
```diff
@@ -27,13 +27,43 @@ const checkBranch = async (context) => {
...
+    await context.github.issues.createComment(commentParams);
+    // Close the pull request.
+    const closeIssueParams = context.repo({
+      issue_number: pullRequest.number,
+      state: 'closed',
+    });
+    await context.github.issues.update(closeIssueParams);
+  }
+
+  else if (branchIsInvalid) {
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:59]: I believe the message for this section needs to be updated as well since we are now handling develop branch changes separately right?
Also, can we have separate `else if` blocks for each of the scenarios as suggested in the issue, https://github.com/oppia/oppiabot/issues/147?

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Thanks, @riyasingh1004!

I took a first pass.

File: lib/checkPullRequestBranch.js (hunk 2 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js:59]: @jameesjohn there are two blocks one for checking develop branch and another for invalid prefix branches. Are you suggesting adding cases for each develop,test and release prefix branches with their separate messages??

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL . I think these messages are better than before as its less ambiguous.

File: lib/checkPullRequestBranch.js [hunk 3]
```diff
@@ -27,13 +27,44 @@ const checkBranch = async (context) => {
     return branchName.startsWith(prefix);
   });
 
-  if (branchIsInvalid) {
+  if(branchName==='develop')
+  {
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:31]:
```suggestion
  if (branchName === 'develop') {
```

File: lib/checkPullRequestBranch.js [hunk 4]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-const DISSALLOWED_BRANCH_PREFIXES = ['develop', 'release-', 'test-'];
+const DISSALLOWED_BRANCH_PREFIXES = ['develop-', 'release-', 'test-'];
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:15]: Why this change?

File: lib/checkPullRequestBranch.js (hunk 2 above)
Comment (jameesjohn) [reply lib/checkPullRequestBranch.js:60]: Yes.
I think that was the aim of the issue, and that will be clearer.

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Thanks, @riyasingh1004. I took another pass.

File: lib/checkPullRequestBranch.js (hunk 4 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js:15]: Because now the invalid names include branches with prefixes. So this differentiates between develop branch and develop prefixed branch.

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL

**Comment by riyasingh1004** [comment]
@jameesjohn  Are there any more changes here??

File: lib/checkPullRequestBranch.js (hunk 2 above)
Comment (jameesjohn) [reply lib/checkPullRequestBranch.js:60]: Has this been fixed?

File: lib/checkPullRequestBranch.js [hunk 5]
```diff
@@ -27,13 +27,44 @@ const checkBranch = async (context) => {
     return branchName.startsWith(prefix);
   });
 
-  if (branchIsInvalid) {
+  if(branchName === 'develop')
+  {
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:31]: The formatting here is not correct.
```suggestion
  if (branchName === 'develop') {
```

**Comment by riyasingh1004** [comment]
@jameesjohn  PTAL

File: lib/checkPullRequestBranch.js (hunk 2 above)
Comment (jameesjohn) [reply lib/checkPullRequestBranch.js:59]: @riyasingh1004 ping.

File: lib/checkPullRequestBranch.js [hunk 6]
```diff
@@ -27,13 +27,43 @@ const checkBranch = async (context) => {
...
+    const commentParams = context.repo({
+      issue_number: pullRequest.number,
+      body:
+        'Hi @' + prAuthor +
+        ', PRs made from develop branch are not allowed. Also branches ' +
+        'whose name is prefixed with develop, release or test are not ' +
+        'allowed. So this PR is being closed. Please make your changes in ' +
+        'another branch and send in the PR. To learn more about contributing ' +
+        'to Oppia, take a look at our ' +
+        wiki + ' (Rule 1 specifically). Thanks!',
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:47]: See https://github.com/oppia/oppiabot/pull/196/files#r572916291

**Comment by jameesjohn** [comment]
Hi @riyasingh1004 any updates on this?

**Comment by riyasingh1004** [comment]
@jameesjohn PTAL

File: lib/checkPullRequestBranch.js [hunk 7]
```diff
@@ -23,28 +22,88 @@ const checkBranch = async (context) => {
...
+        'Hi @' + prAuthor +
+        ', PRs made from a branch whose name is prefixed with release ' +
+        'are not allowed. Also PRs made from develop branch or from a ' +
+        'branch whose name is prefixed with develop or test are not '+
+        'allowed. So this PR is being closed. Please make your changes in ' +
+        'another branch and send in the PR. To learn more about contributing ' +
+        'to Oppia, take a look at our ' +
+        wiki + ' (Rule 1 specifically). Thanks!',
+    });
+
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:106]: So the point of splitting this into different cases is to have different messages for them. 
For example, if the branch starts with `test`, the message should be `Hi  @author, PRs made from branch whose name is prefixed with test is not allowed. So this PR is being closed...`
And similar for other prefixes.

File: lib/checkPullRequestBranch.js [hunk 8]
```diff
@@ -23,28 +22,88 @@ const checkBranch = async (context) => {
...
       body:
         'Hi @' + prAuthor +
-        ', PRs made from develop branch or from a branch ' +
+        ', PRs made from develop branch are not allowed. Also branches ' +
         'whose name is prefixed with develop, release or test are not ' +
         'allowed. So this PR is being closed. Please make your changes in ' +
         'another branch and send in the PR. To learn more about contributing ' +
         'to Oppia, take a look at our ' +
         wiki + ' (Rule 1 specifically). Thanks!',
     });
```
Comment (jameesjohn) [inline lib/checkPullRequestBranch.js:45]: This can be refactored to store the message for each case in a messages variable, and then make the request to comment on the PR at the end.

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Thanks, @riyasingh1004. Few comments, PTAL.

**Comment by riyasingh1004** [comment]
@jameesjohn  PTAL

**Review by jameesjohn — APPROVED** [review]
LGTM

**Comment by jameesjohn** [comment]
PTAL @vojtechjelinek

File: lib/checkPullRequestBranch.js [hunk 9]
```diff
@@ -23,28 +22,62 @@ const checkBranch = async (context) => {
...
+  message = '';
+  invalid_branch_check = false;
+  
+
+  if(branchName === 'develop'){
+    
+    invalid_branch_check = true;
+    message = ', PRs made from develop branch are not allowed. Also branches '+
+              'whose name is prefixed with develop, release or test are not allowed.';
+  }
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:35]:
```suggestion
  if (branchName === 'develop') {
    invalid_branch_check = true;
    message = (
      ', PRs made from develop branch are not allowed. Also branches ' +
      'whose name is prefixed with develop, release or test are not allowed.'
    );
  }
```

File: lib/checkPullRequestBranch.js [hunk 10]
```diff
@@ -23,28 +22,62 @@ const checkBranch = async (context) => {
...
+
+  else if (branchName.startsWith('test')) {
+
+    invalid_branch_check = true;
+    message = ', PRs made from a branch whose name is prefixed with test ' +
+              'are not allowed. Also PRs made from develop branch or from a branch '+
+              'whose name is prefixed with develop or release are not allowed.';
+
+  }
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:60]: This can be simplified right? I would not mention the second part `Also PRs made from develop branch or from a branch whose name is prefixed with develop or test are not allowed.` an just iterate through the disallowed prefixes and if it is prefixed with it do `', PRs made from a branch whose name is prefixed with ${prefix} are not allowed.'`

File: lib/checkPullRequestBranch.js [hunk 11]
```diff
@@ -23,28 +22,62 @@ const checkBranch = async (context) => {
...
+
+  if(invalid_branch_check === true)
+  {
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:63]:
```suggestion
  if (invalid_branch_check) {
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Left a fe comments.

**Comment by riyasingh1004** [comment]
@vojtechjelinek I tried optimising the message but since the message is returned at last for all the cases, I just changed the first line for all cases. Because for develop branch check and prefixed branch check the fist line of the message is different. So instead of making two return values changing the first line provided better optimization.

**Comment by vojtechjelinek** [comment]
@riyasingh1004 Please respond to all my comments individually, if you've fixed them you can use jsut "Done.". Thanks.

File: lib/checkPullRequestBranch.js (hunk 9 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: @vojtechjelinek Is this still required, as the message size is reduced??

File: lib/checkPullRequestBranch.js (hunk 10 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

File: lib/checkPullRequestBranch.js (hunk 11 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

File: lib/checkPullRequestBranch.js [hunk 12]
```diff
@@ -23,28 +22,60 @@ const checkBranch = async (context) => {
...
+
+  else if (branchName.startsWith('test')) {
+
+    invalid_branch_check = true;
+    message = ', PRs made from a branch whose name is prefixed with test are not allowed.';
+              
+  }
+
+  if(invalid_branch_check)
+  {
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:59]:
```suggestion
  if (invalid_branch_check) {
```

File: lib/checkPullRequestBranch.js [hunk 13]
```diff
@@ -23,28 +22,60 @@ const checkBranch = async (context) => {
...
+  message = '';
+  invalid_branch_check = false;
+  
+
+  if(branchName === 'develop'){
+    
+    invalid_branch_check = true;
+    message = ', PRs made from develop branch are not allowed.'
+    
+  }
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:35]:
```suggestion
  if (branchName === 'develop') {
    invalid_branch_check = true;
    message = ', PRs made from develop branch are not allowed.'
  }
```

File: lib/checkPullRequestBranch.js [hunk 14]
```diff
@@ -23,28 +22,60 @@ const checkBranch = async (context) => {
...
   const branchName = pullRequest.head.ref;
-  const branchIsInvalid = DISSALLOWED_BRANCH_PREFIXES.some((prefix) => {
-    return branchName.startsWith(prefix);
-  });
 
-  if (branchIsInvalid) {
+  message = '';
+  invalid_branch_check = false;
+  
+
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:29]: Use just one empty line.

File: lib/checkPullRequestBranch.js [hunk 15]
```diff
@@ -23,28 +22,60 @@ const checkBranch = async (context) => {
...
+    message = ', PRs made from a branch whose name is prefixed with release are not allowed.';
+              
+  }
+
+  else if (branchName.startsWith('test')) {
+
+    invalid_branch_check = true;
+    message = ', PRs made from a branch whose name is prefixed with test are not allowed.';
+              
+  }
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:56]: Here you can do this instead a shorten the code.
```suggestion
  else {
    invalid_prefixes = ['develop', 'release', 'test']
    for (i in invalid_prefixes) {
      if (branchName.startsWith(invalid_prefixes[i])) {
        invalid_branch_check = true;
        message = (
          `, PRs made from a branch whose name is prefixed with `
          `${invalid_prefixes[i]} are not allowed.`
        );            
      }
    }
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Left a few more comments.

**Comment by riyasingh1004** [comment]
@vojtechjelinek  PTAL

File: lib/checkPullRequestBranch.js (hunk 13 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

File: lib/checkPullRequestBranch.js (hunk 14 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

**Comment by riyasingh1004** [comment]
@vojtechjelinek Sorry about earlier. I have addressed all the comments individually. I hope nothing was left.

File: lib/checkPullRequestBranch.js [hunk 16]
```diff
@@ -23,28 +22,49 @@ const checkBranch = async (context) => {
...
+  invalid_branch_check = false;
+  
+  if(branchName === 'develop'){
+     invalid_branch_check = true;
+     message = ', PRs made from develop branch are not allowed.';
+  }
+
+  else {
+    invalid_prefixes = ['develop', 'release', 'test'];
+    for(i=0;i<3;i++) {
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:37]: Why not?
```suggestion
    for (let i in invalid_prefixes) {
```

File: lib/checkPullRequestBranch.js [hunk 17]
```diff
@@ -23,28 +22,49 @@ const checkBranch = async (context) => {
...
+ 
+  message = '';
+  invalid_branch_check = false;
+  
+  if(branchName === 'develop'){
+     invalid_branch_check = true;
+     message = ', PRs made from develop branch are not allowed.';
+  }
+
+  else {
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:35]:
```suggestion
  } else {
```

File: lib/checkPullRequestBranch.js [hunk 18]
```diff
@@ -23,28 +22,49 @@ const checkBranch = async (context) => {
...
+        invalid_branch_check = true;
+        message = (
+          ', PRs made from a branch whose name is prefixed with ' +
+          invalid_prefixes[i] + ' are not allowed.'
+        );
+      }
+    } 
+  }
+
+  if(invalid_branch_check){
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:48]:
```suggestion
  if (invalid_branch_check) {
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Few other comments.

File: lib/checkPullRequestBranch.js (hunk 16 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

File: lib/checkPullRequestBranch.js (hunk 18 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

**Comment by riyasingh1004** [comment]
@vojtechjelinek  PTAL

File: lib/checkPullRequestBranch.js [hunk 19]
```diff
@@ -23,28 +22,47 @@ const checkBranch = async (context) => {
...
+  message = '';
+  invalid_branch_check = false;
+  
+  if (branchName === 'develop') {
+     invalid_branch_check = true;
+     message = ', PRs made from develop branch are not allowed.';
+  } else {
+    invalid_prefixes = ['develop', 'release', 'test'];
+    for (let i in invalid_prefixes) {
+      if(branchName.startsWith(invalid_prefixes[i])) {
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:36]:
```suggestion
      if (branchName.startsWith(invalid_prefixes[i])) {
```

File: lib/checkPullRequestBranch.js [hunk 20]
```diff
@@ -23,28 +22,47 @@ const checkBranch = async (context) => {
...
-  });
 
-  if (branchIsInvalid) {
+ 
+  message = '';
+  invalid_branch_check = false;
+  
+  if (branchName === 'develop') {
+     invalid_branch_check = true;
+     message = ', PRs made from develop branch are not allowed.';
```
Comment (vojtechjelinek) [inline lib/checkPullRequestBranch.js:32]:
```suggestion
    invalid_branch_check = true;
    message = ', PRs made from develop branch are not allowed.';
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Two last remaining comments.

File: lib/checkPullRequestBranch.js (hunk 19 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

File: lib/checkPullRequestBranch.js (hunk 20 above)
Comment (riyasingh1004) [reply lib/checkPullRequestBranch.js]: Done

**Comment by riyasingh1004** [comment]
@vojtechjelinek  PTAL

**Review by vojtechjelinek — APPROVED** [review]
LGTM!
